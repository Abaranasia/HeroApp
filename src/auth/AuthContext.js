import { createContext } from "react";

export const AuthContext= createContext();
// lleva may√∫suclas porque es un componente, en este caso un high order component (apertura y cierre)