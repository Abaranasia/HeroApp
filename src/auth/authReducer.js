import { types } from "../types/types";

/** Reducer para tener bajo los estados de autenticación en la app
 * Toma los estados (login y logout) desde types
 * Este reducer estará compartido en toda la aplicación gracias a un context (AuthContext)
*/

export const authReducer = (state= {}, action) => {
    switch (action.type) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }
        case types.logout:
            return {
                logged: false
            }
        default:
            return state;
    }
}