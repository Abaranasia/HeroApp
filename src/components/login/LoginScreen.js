import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext'
import { types } from '../../types/types'

export const LoginScreen = ( { history }) => {

const { dispatch } = useContext(AuthContext)
    
    const handleLogin = () => {
        //history.push('/') // es bien, pero depende de que se haya visitado
        

        const lastpath = localStorage.getItem('lastpath') || '/'; // Busca un path de una visita anterior o lleva a '/'
        const action = { // defino la acción de añadir
            type: types.login,
            payload: { name:'Ran' }
        };
    
        dispatch (action) // enviamos la acción al reducer

        history.replace(lastpath) 
        /** replace borra el rastro y no será navegable de vuelta con el botón 
         * atrás del navegador, aunque aquí carga, si hay, un path anterior */
    }

    return (
        <div className="container mt-5">
            <h2>Login</h2>
            <hr/>
            <button
                className="btn btn-primary"
                onClick= {handleLogin}
            >
                Login
            </button>
        </div>
    )
}
