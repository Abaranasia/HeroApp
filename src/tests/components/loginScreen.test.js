import { mount } from 'enzyme';
import { AuthContext } from '../../auth/AuthContext';

import { LoginScreen } from '../../components/login/LoginScreen';
import { types } from '../../types/types';


describe('Debe mostrarse correctamente', () => {
    /** Mostrarse, handleLogin con replace y localStorage; ojo con history */

    const historyMock = {
        replace: jest.fn(),
    }

    const contextValue = { // valores necesarios para enviar en el context provider
        dispatch: jest.fn(),
    }

    const wrapper = mount(
        <AuthContext.Provider value= {contextValue}>
                <LoginScreen history= { historyMock } />
        </AuthContext.Provider>
    );

    test('Debe mostrarse correctamente', () => {

        expect (wrapper).toMatchSnapshot();
        expect(wrapper.find('h2').text().trim()).toBe('Login')
    });
    
    test('Debe realizar el dispatch y la navegación en handleLogin', () => {
        
        const action = { // defino la acción de añadir
            type: types.login,
            payload: { name:'Ran' }
        };

        //const handleClick = wrapper.find('button').prop('onClick');
        const handleClick = () => {wrapper.find('button').simulate('click')}
        
        /** hacemos esta asignación para poder invocar el click varias veces:
         * una sin localStorage previo y otra con una ruta guardada
         * wrapper.find('button').simulate('click') no se puede usar como funcion pues devuelve un objeto
         */
       // console.log(handleClick.html());
        handleClick();
        
        expect(contextValue.dispatch).toHaveBeenCalledWith(action);
        expect(historyMock.replace).toHaveBeenCalledWith('/')
        //LLamada sin localStorage previo

        const lastpath= '/marvel';
        localStorage.setItem('lastpath', lastpath);
        //guardamos en localStorage
        
        handleClick();
        
        expect(historyMock.replace).toHaveBeenCalledWith('/marvel')
        //Probamos si leyó del localStorage

        
    });


});