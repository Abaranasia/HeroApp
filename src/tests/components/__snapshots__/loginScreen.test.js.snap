// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Debe mostrarse correctamente Debe mostrarse correctamente 1`] = `
<div
  className="container mt-5"
>
  <h2>
    Login
  </h2>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
